## Diabetes AI Project

2023/01/17
1726i

lets code
lets get init

# import data

In [13]:
import csv
import numpy as np
import torch


f =  open("C:/Users/OWO/Documents/AA_CODE/@Projects/Projects/ai_0003_lc_diabetes/diabetes.csv")
read = csv.reader(f)

#data to list
data = []
for row in read:
  data.append(row)

#changing str to float
del data[0]
for column in range(len(data)):
  for row in range(len(data[0])):
    data[column][row] = float(data[column][row])
  

x_train_data = torch.Tensor(np.array(data)[:718,:-1])
y_train_data = torch.Tensor(np.array(data)[:718,-1:])

"""
print(x_train_data.tolist())
print("\n\n\n")
print(y_train_data.tolist())
"""
print(len(x_train_data))
print(len(y_train_data))
f.close()

718
718


# Hardware check

In [14]:
##Hardwere
import torch
if torch.cuda.is_available() == True:
  device = 'cuda'
  templist = [1,2,3]
  templist = torch.FloatTensor(templist).to(device)
  print("Cude torch working : ",end="")
  print(templist.is_cuda)
  print("current device no. : ",end="")
  print(torch.cuda.current_device())
  print("GPU device count : ",end="")
  print(torch.cuda.device_count())
  print("GPU name : ",end="")
  print(torch.cuda.get_device_name(0))
  print("device : ",device)
else:
  print("cant use gpu , activating cpu")
  device = 'cpu'

Cude torch working : True
current device no. : 0
GPU device count : 1
GPU name : NVIDIA GeForce GTX 1080
device :  cuda


# ML

In [18]:
#original

import torch
import torch.optim as optim
import torch.nn as nn

#Hardware
print("GPU Status ",end="")
print("Working Well") if torch.cuda.is_available() else print("Not good, Activating CPU calculation")
device = 'cuda' if torch.cuda.is_available() else 'cpu'

x_train_data = x_train_data.to(device)
y_train_data = y_train_data.to(device)

#lr_list = [1,1e-2,1e-3,1e-4,1e-5,1e-6,1e-7,1e-8,1e-9,1e-10,1e-11,1e-12,1e-13,1e-14,1e-15]
lr_list = [1e-4]
for LR in lr_list:

  W = torch.zeros([8,1],requires_grad=True,device=device)
  b = torch.zeros(1,requires_grad=True,device=device)
  
  optimizer = optim.SGD([W],lr = LR)
  nb_epochs = 1000
  
  for epoch in range(nb_epochs + 1):
    #H(x)
    hypothesis = torch.sigmoid(torch.matmul(x_train_data,W) + b)
    
    #cost = -(((y_train_data * torch.log(hypothesis))) + ((1-y_train_data)*torch.log(1-hypothesis))).mean() #nan problem
    loss = torch.nn.BCELoss()
    cost = loss(hypothesis,y_train_data)
    optimizer.zero_grad()
    cost.backward()
    optimizer.step()
  
#result
  print("LR : %E cost : %f"%(LR,cost.item()))
  if torch.isfinite(cost):#if its not nan
    #print(W.tolist(),b.tolist())
    continue


GPU Status Working Well
LR : 1.000000E-04 cost : 0.617943


# Many LR and results
LR : 1.000000E+00 cost : 34.679665

LR : 1.000000E-02 cost : 33.583466

LR : 1.000000E-03 cost : 3.829255

LR : 1.000000E-04 cost : 0.617943

LR : 1.000000E-05 cost : 0.628474

LR : 1.000000E-06 cost : 0.658451

LR : 1.000000E-07 cost : 0.677100

LR : 1.000000E-08 cost : 0.690457

LR : 1.000000E-09 cost : 0.692861

LR : 1.000000E-10 cost : 0.693118

LR : 1.000000E-11 cost : 0.693144

LR : 1.000000E-12 cost : 0.693147

LR : 1.000000E-13 cost : 0.693147

LR : 1.000000E-14 cost : 0.693147

LR : 1.000000E-15 cost : 0.693147



## Evaluation

In [16]:
#Getting the rest data 720~769
import csv
import numpy as np
import torch

f =  open("C:/Users/OWO/Documents/AA_CODE/@Projects/Projects/ai_0003_lc_diabetes/diabetes.csv")
read = csv.reader(f)

#data to list
data = []
for row in read:
  data.append(row)

#changing str to float
del data[0]
for column in range(len(data)):
  for row in range(len(data[0])):
    data[column][row] = float(data[column][row])


device = 'cuda' if torch.cuda.is_available() else 'cpu'


x_test_data = torch.Tensor(np.array(data)[718:,:-1]).to(device)
y_test_data = torch.Tensor(np.array(data)[718:,-1:]).to(device)

#print(x_test_data.tolist())
#print(y_test_data)
#print(len(x_test_data))
#print(len(y_test_data))

f.close()
threshold = [0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9]
for num in threshold:
  hypothesis = torch.sigmoid(torch.matmul(x_test_data,W)+b)#.to(device)#hypothesis
  #print(hypothesis)
  prediction = hypothesis >= torch.FloatTensor([num]).to(device)#making it to binary // prediction will be in true or false
  correct_prediction = (prediction.float() == y_test_data) #if it matches it will be true  .float() will turn true to 1
  print("threshold : ",num,"accuracy : ",correct_prediction.sum().item()/len(correct_prediction)*100,"%")



threshold :  0.1 accuracy :  38.0 %
threshold :  0.2 accuracy :  42.0 %
threshold :  0.3 accuracy :  50.0 %
threshold :  0.4 accuracy :  68.0 %
threshold :  0.5 accuracy :  66.0 %
threshold :  0.6 accuracy :  62.0 %
threshold :  0.7 accuracy :  62.0 %
threshold :  0.8 accuracy :  62.0 %
threshold :  0.9 accuracy :  62.0 %


# Result 68% accuracy

# End


2023/01/18 1354i
I am kind of disappointed that my accuracy is only 64%.
But I learned a lot in this project.

the data source : https://www.kaggle.com/code/mathchi/diagnostic-a-patient-has-diabetes